File: LeetCode/Array/Rust/Leetcode498-diagonal-traverse.rs
impl Solution {
    pub fn find_diagonal_order(mat: Vec<Vec<i32>>) -> Vec<i32> {
        let (n, m, mut ans) = (mat.len() as i32, mat[0].len() as i32, Vec::new());
        for i in 0..(n + m - 1) {
            if i % 2 == 1 {
                let (mut x, mut y) = (0, i);
                if i >= m {
                    x = i - m + 1;
                    y = m - 1;
                }
                while x < n && y >= 0 {
                    // println!("{},{}", x, y);
                    ans.push(mat[x as usize][y as usize]);
                    x += 1;
                    y -= 1;
                }
            } else {
                let (mut x, mut y) = (i, 0);
                if x >= n {
                    x = n - 1;
                    y = i - n + 1;
                }
                while x >= 0 && y < m {
                    ans.push(mat[x as usize][y as usize]);
                    x -= 1;
                    y += 1;
                }
            }
        }
        ans
    }
}

File: LeetCode/Array/Python/Leetcode498-diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        ans, m, n = [], len(mat[0]), len(mat)
        for i in range(n + m - 1):
            if i % 2 == 1:
                x = 0 if i < m else i - m + 1
                y = i if i < m else m - 1
                while x < n and y >= 0:
                    ans.append(mat[x][y])
                    x += 1
                    y -= 1
            else:
                x = i if i < n else n - 1
                y = 0 if i < n else i - n + 1
                while x >= 0 and y < m:
                    ans.append(mat[x][y])
                    x -= 1
                    y += 1
        return ans
